src/core/json/path.cc

```diff
@@ -0,0 +1,223 @@
+// Copyright 2024, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "src/core/json/path.h"
+
+#include <absl/types/span.h>
+
+#include "base/expected.hpp"
+#include "base/logging.h"
+#include "src/core/json_object.h"
+
+using namespace std;
+using nonstd::make_unexpected;
+
+namespace dfly::json {
+
+namespace {
+
+class Dfs {
+ public:
+  using Cb = std::function<void(const JsonType&)>;
+
+  // TODO: for some operations we need to know the type of mismatches.
+  void Traverse(absl::Span<const PathSegment> path, const JsonType& json, const Cb& callback);
+
+  unsigned matches() const {
+    return matches_;
+  }
+
+ private:
+  enum MatchStatus {
+    OUT_OF_BOUNDS,
+    MISMATCH,
+  };
+
+  bool TraverseImpl(absl::Span<const PathSegment> path, const Cb& callback);
+
+  nonstd::expected<void, MatchStatus> PerformStep(const PathSegment& segment, const JsonType& node,
+                                                  const Cb& callback);
+
+  void DoCall(const Cb& callback, const JsonType& node) {
+    ++matches_;
+    callback(node);
+  }
+
+  using AdvanceResult = nonstd::expected<const JsonType*, MatchStatus>;
+
+  struct Item {
+    const JsonType* obj;
+    variant<monostate, JsonType::const_object_iterator, JsonType::const_array_iterator> state;
+    Item(const JsonType* o) : obj(o) {
+    }
+
+    // Returns the next object to traverse
+    // or null if traverse was exhausted or the segment does not match.
+    AdvanceResult Advance(const PathSegment& segment);
+  };
+
+  unsigned matches_ = 0;
+};
+
+auto Dfs::Item::Advance(const PathSegment& segment) -> AdvanceResult {
+  if (std::holds_alternative<monostate>(state)) {
```

Comment:
too many include blocks, I suggest to move switch out of IF block

Reply:
Fixed